File: app/controllers/iop/transfer/impl/ExternalReservationHandlerImpl.java
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl;

import com.fasterxml.jackson.databind.JsonNode;
import controllers.iop.transfer.api.ExternalReservationHandler;
import impl.mail.EmailComposer;
import io.ebean.DB;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.function.Function;
import javax.inject.Inject;
import miscellaneous.config.ConfigReader;
import miscellaneous.datetime.DateTimeHandler;
import models.enrolment.ExamEnrolment;
import models.enrolment.ExternalReservation;
import models.enrolment.Reservation;
import models.user.User;
import org.apache.pekko.actor.ActorSystem;
import org.joda.time.DateTime;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import play.libs.ws.WSClient;
import play.libs.ws.WSRequest;
import play.libs.ws.WSResponse;
import play.mvc.Http;
import play.mvc.Result;
import play.mvc.Results;
import scala.concurrent.duration.Duration;
import scala.jdk.javaapi.OptionConverters;

public class ExternalReservationHandlerImpl implements ExternalReservationHandler {

    @Inject
    WSClient wsClient;

    @Inject
    ActorSystem system;

    @Inject
    EmailComposer emailComposer;

    @Inject
    DateTimeHandler dateTimeHandler;

    @Inject
    ConfigReader configReader;

    private final Logger logger = LoggerFactory.getLogger(ExternalReservationHandlerImpl.class);

    private URL parseUrl(String orgRef, String facilityRef, String reservationRef) throws MalformedURLException {
        StringBuilder sb = new StringBuilder(configReader.getIopHost());
        sb.append(String.format("/api/organisations/%s/facilities/%s/reservations", orgRef, facilityRef));
        if (reservationRef != null) {
            sb.append("/").append(reservationRef);
        }
        return URI.create(sb.toString()).toURL();
    }

    @Override
    public CompletionStage<Optional<Integer>> removeExternalReservation(Reservation reservation) {
        ExternalReservation external = reservation.getExternalReservation();
        URL url;
        try {
            url = parseUrl(external.getOrgRef(), external.getRoomRef(), reservation.getExternalRef());
        } catch (MalformedURLException e) {
            return CompletableFuture.completedFuture(Optional.of(Http.Status.INTERNAL_SERVER_ERROR));
        }
        WSRequest request = wsClient.url(url.toString());
        Function<WSResponse, Optional<Integer>> onSuccess = response -> {
            if (response.getStatus() != Http.Status.OK) {
                return Optional.of(Http.Status.INTERNAL_SERVER_ERROR);
            }
            return Optional.empty();
        };
        return request.delete().thenApplyAsync(onSuccess);
    }

    private CompletionStage<Result> requestRemoval(String ref, User user, String msg) throws IOException {
        final ExamEnrolment enrolment = DB.find(ExamEnrolment.class)
            .fetch("reservation")
            .fetch("reservation.machine")
            .fetch("reservation.machine.room")
            .where()
            .eq("user.id", user.getId())
            .eq("reservation.externalRef", ref)
            .findOne();
        if (enrolment == null) {
            return CompletableFuture.completedFuture(
                Results.notFound(String.format("No reservation with ref %s for current user.", ref))
            );
        }
        // Removal not permitted if reservation is in the past or ongoing
        final Reservation reservation = enrolment.getReservation();
        DateTime now = dateTimeHandler.adjustDST(DateTime.now(), reservation.getExternalReservation());
        if (reservation.toInterval().isBefore(now) || reservation.toInterval().contains(now)) {
            return CompletableFuture.completedFuture(Results.forbidden("i18n_reservation_in_effect"));
        }
        // good to go
        ExternalReservation external = reservation.getExternalReservation();
        URL url = parseUrl(external.getOrgRef(), external.getRoomRef(), ref);
        WSRequest request = wsClient.url(url.toString());
        Function<WSResponse, Result> onSuccess = response -> {
            if (response.getStatus() != Http.Status.OK) {
                JsonNode root = response.asJson();
                return Results.internalServerError(root.get("message").asText("Connection refused"));
            }
            enrolment.setReservation(null);
            enrolment.setReservationCanceled(true);
            DB.save(enrolment);
            reservation.delete();

            // send email asynchronously
            boolean isStudentUser = user.equals(enrolment.getUser());
            system
                .scheduler()
                .scheduleOnce(
                    Duration.create(1, TimeUnit.SECONDS),
                    () -> {
                        emailComposer.composeReservationCancellationNotification(
                            enrolment.getUser(),
                            reservation,
                            OptionConverters.toScala(Optional.of(msg)),
                            isStudentUser,
                            enrolment
                        );
                        logger.info("Reservation cancellation confirmation email sent");
                    },
                    system.dispatcher()
                );

            return Results.ok();
        };
        return request.delete().thenApplyAsync(onSuccess);
    }

    // remove reservation on external side, initiated by reservation holder
    @Override
    public CompletionStage<Result> removeReservation(Reservation reservation, User user, String msg) {
        if (reservation.getExternalReservation() == null) {
            return CompletableFuture.completedFuture(Results.ok());
        }
        try {
            return requestRemoval(reservation.getExternalRef(), user, msg);
        } catch (IOException e) {
            return CompletableFuture.completedFuture(Results.internalServerError(e.getMessage()));
        }
    }
}

File: app/controllers/iop/transfer/impl/ExternalReservationHandlerImpl.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package controllers.iop.transfer.impl

import controllers.iop.transfer.api.ExternalReservationHandler
import impl.mail.EmailComposer
import io.ebean.DB
import miscellaneous.config.ConfigReader
import miscellaneous.datetime.DateTimeHandler
import models.enrolment.{ExamEnrolment, Reservation}
import models.user.User
import org.apache.pekko.actor.ActorSystem
import org.joda.time.DateTime
import play.api.Logging
import play.api.http.Status.*
import play.api.libs.ws.{JsonBodyWritables, WSClient}
import play.api.mvc.Result
import play.api.mvc.Results.*

import java.net.{MalformedURLException, URI}
import javax.inject.Inject
import scala.concurrent.{ExecutionContext, Future}
import scala.concurrent.duration.*

class ExternalReservationHandlerImpl @Inject() (
    wsClient: WSClient,
    system: ActorSystem,
    emailComposer: EmailComposer,
    dateTimeHandler: DateTimeHandler,
    configReader: ConfigReader
)(implicit ec: ExecutionContext)
    extends ExternalReservationHandler
    with JsonBodyWritables
    with Logging:

  private def parseUrl(orgRef: String, facilityRef: String, reservationRef: Option[String]): URI =
    val sb = new StringBuilder(configReader.getIopHost)
    sb.append(s"/api/organisations/$orgRef/facilities/$facilityRef/reservations")
    reservationRef.foreach(ref => sb.append(s"/$ref"))
    URI.create(sb.toString)

  override def removeExternalReservation(reservation: Reservation): Future[Option[Int]] =
    val external = reservation.getExternalReservation
    try
      val url = parseUrl(external.getOrgRef, external.getRoomRef, Some(reservation.getExternalRef))
      wsClient
        .url(url.toString)
        .delete()
        .map { response =>
          if response.status != OK then Some(INTERNAL_SERVER_ERROR)
          else None
        }
        .recover { case _ =>
          Some(INTERNAL_SERVER_ERROR)
        }
    catch
      case _: MalformedURLException =>
        Future.successful(Some(INTERNAL_SERVER_ERROR))

  private def requestRemoval(ref: String, user: User, msg: String): Future[Result] =
    val enrolmentOpt = Option(
      DB.find(classOf[ExamEnrolment])
        .fetch("reservation")
        .fetch("reservation.machine")
        .fetch("reservation.machine.room")
        .where()
        .eq("user.id", user.getId)
        .eq("reservation.externalRef", ref)
        .findOne()
    )

    enrolmentOpt match
      case None =>
        Future.successful(NotFound(s"No reservation with ref $ref for current user."))
      case Some(enrolment) =>
        // Removal is not permitted if the reservation is in the past or ongoing
        val reservation = enrolment.getReservation
        val now         = dateTimeHandler.adjustDST(DateTime.now(), reservation.getExternalReservation)

        if reservation.toInterval.isBefore(now) || reservation.toInterval.contains(now) then
          Future.successful(Forbidden("i18n_reservation_in_effect"))
        else
          // good to go
          val external = reservation.getExternalReservation
          try
            val url = parseUrl(external.getOrgRef, external.getRoomRef, Some(ref))
            wsClient
              .url(url.toString)
              .delete()
              .map { response =>
                if response.status != OK then
                  val msg = (response.json \ "message").asOpt[String].getOrElse("Connection refused")
                  InternalServerError(msg)
                else
                  enrolment.setReservation(null)
                  enrolment.setReservationCanceled(true)
                  DB.save(enrolment)
                  reservation.delete()

                  // send email asynchronously
                  val isStudentUser = user == enrolment.getUser
                  system.scheduler.scheduleOnce(1.second) {
                    emailComposer.composeReservationCancellationNotification(
                      enrolment.getUser,
                      reservation,
                      Some(msg),
                      isStudentUser,
                      enrolment
                    )
                    logger.info("Reservation cancellation confirmation email sent")
                  }(using system.dispatcher)

                  Ok
              }
          catch
            case e: MalformedURLException =>
              Future.successful(InternalServerError(e.getMessage))

  // remove reservation on the external side, initiated by the reservation holder
  override def removeReservation(reservation: Reservation, user: User, msg: String): Future[Result] =
    if Option(reservation.getExternalReservation).isEmpty then Future.successful(Ok)
    else requestRemoval(reservation.getExternalRef, user, msg)
